File: src/test/java/no/nav/bidrag/beregn/felles/FellesPeriodeTest.java
package no.nav.bidrag.beregn.felles;

import static org.assertj.core.api.Assertions.assertThat;
import static org.junit.jupiter.api.Assertions.assertAll;

import java.time.LocalDate;
import java.util.ArrayList;
import java.util.List;
import no.nav.bidrag.beregn.felles.bo.Periode;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;

public class FellesPeriodeTest extends FellesPeriode {

  @Test
  @DisplayName("Test av periodisering. Periodene i grunnlaget skal gjenspeiles i resultatperiodene")
  void testPeriodisering() {

    List<Periode> periodeListeMedJustering = new ArrayList<>();
    periodeListeMedJustering.add(new Periode(LocalDate.parse("2017-01-01"), LocalDate.parse("2018-01-01")));
    periodeListeMedJustering.add(new Periode(LocalDate.parse("2018-01-01"), LocalDate.parse("2019-01-01")));
    periodeListeMedJustering.add(new Periode(LocalDate.parse("2019-01-01"), LocalDate.parse("2020-01-01")));
    periodeListeMedJustering.add(new Periode(LocalDate.parse("2019-01-01"), null));

    List<Periode> periodeListeUtenJustering = new ArrayList<>();
    periodeListeUtenJustering.add(new Periode(LocalDate.parse("2017-01-01"), LocalDate.parse("2018-01-01")));
    periodeListeUtenJustering.add(new Periode(LocalDate.parse("2018-01-01"), LocalDate.parse("2019-01-01")));
    periodeListeUtenJustering.add(new Periode(LocalDate.parse("2019-01-01"), LocalDate.parse("2020-01-01")));

    mergeSluttperiode(periodeListeMedJustering, LocalDate.parse("2020-01-01"));
    mergeSluttperiode(periodeListeUtenJustering, LocalDate.parse("2020-01-01"));

    assertAll(
        () -> assertThat(periodeListeMedJustering).isNotNull(),
        () -> assertThat(periodeListeMedJustering.size()).isEqualTo(3),
        () -> assertThat(periodeListeMedJustering.get(0).getDatoFom()).isEqualTo(LocalDate.parse("2017-01-01")),
        () -> assertThat(periodeListeMedJustering.get(0).getDatoTil()).isEqualTo(LocalDate.parse("2018-01-01")),
        () -> assertThat(periodeListeMedJustering.get(1).getDatoFom()).isEqualTo(LocalDate.parse("2018-01-01")),
        () -> assertThat(periodeListeMedJustering.get(1).getDatoTil()).isEqualTo(LocalDate.parse("2019-01-01")),
        () -> assertThat(periodeListeMedJustering.get(2).getDatoFom()).isEqualTo(LocalDate.parse("2019-01-01")),
        () -> assertThat(periodeListeMedJustering.get(2).getDatoTil()).isNull(),

        () -> assertThat(periodeListeUtenJustering).isNotNull(),
        () -> assertThat(periodeListeUtenJustering.size()).isEqualTo(3),
        () -> assertThat(periodeListeUtenJustering.get(0).getDatoFom()).isEqualTo(LocalDate.parse("2017-01-01")),
        () -> assertThat(periodeListeUtenJustering.get(0).getDatoTil()).isEqualTo(LocalDate.parse("2018-01-01")),
        () -> assertThat(periodeListeUtenJustering.get(1).getDatoFom()).isEqualTo(LocalDate.parse("2018-01-01")),
        () -> assertThat(periodeListeUtenJustering.get(1).getDatoTil()).isEqualTo(LocalDate.parse("2019-01-01")),
        () -> assertThat(periodeListeUtenJustering.get(2).getDatoFom()).isEqualTo(LocalDate.parse("2019-01-01")),
        () -> assertThat(periodeListeUtenJustering.get(2).getDatoTil()).isEqualTo(LocalDate.parse("2020-01-01"))
    );
  }
}

File: src/test/kotlin/no/nav/bidrag/beregn/felles/FellesPeriodeTest.kt
package no.nav.bidrag.beregn.felles

import no.nav.bidrag.beregn.felles.bo.Periode
import org.assertj.core.api.Assertions
import org.junit.jupiter.api.Assertions.assertAll
import org.junit.jupiter.api.DisplayName
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.function.Executable
import java.time.LocalDate

internal class FellesPeriodeTest : FellesPeriode() {

    @Test
    @DisplayName("Test av periodisering. Periodene i grunnlaget skal gjenspeiles i resultatperiodene")
    fun testPeriodisering() {
        val periodeListeMedJustering = mutableListOf(
            Periode(LocalDate.parse("2017-01-01"), LocalDate.parse("2018-01-01")),
            Periode(LocalDate.parse("2018-01-01"), LocalDate.parse("2019-01-01")),
            Periode(LocalDate.parse("2019-01-01"), LocalDate.parse("2020-01-01")),
            Periode(LocalDate.parse("2019-01-01"), null)
        )
        val periodeListeUtenJustering = mutableListOf(
            Periode(LocalDate.parse("2017-01-01"), LocalDate.parse("2018-01-01")),
            Periode(LocalDate.parse("2018-01-01"), LocalDate.parse("2019-01-01")),
            Periode(LocalDate.parse("2019-01-01"), LocalDate.parse("2020-01-01"))
        )

        mergeSluttperiode(periodeListe = periodeListeMedJustering, datoTil = LocalDate.parse("2020-01-01"))
        mergeSluttperiode(periodeListe = periodeListeUtenJustering, datoTil = LocalDate.parse("2020-01-01"))

        assertAll(
            Executable { Assertions.assertThat(periodeListeMedJustering).isNotNull() },
            Executable { Assertions.assertThat(periodeListeMedJustering.size).isEqualTo(3) },
            Executable { Assertions.assertThat(periodeListeMedJustering[0].datoFom).isEqualTo(LocalDate.parse("2017-01-01")) },
            Executable { Assertions.assertThat(periodeListeMedJustering[0].datoTil).isEqualTo(LocalDate.parse("2018-01-01")) },
            Executable { Assertions.assertThat(periodeListeMedJustering[1].datoFom).isEqualTo(LocalDate.parse("2018-01-01")) },
            Executable { Assertions.assertThat(periodeListeMedJustering[1].datoTil).isEqualTo(LocalDate.parse("2019-01-01")) },
            Executable { Assertions.assertThat(periodeListeMedJustering[2].datoFom).isEqualTo(LocalDate.parse("2019-01-01")) },
            Executable { Assertions.assertThat(periodeListeMedJustering[2].datoTil).isNull() },

            Executable { Assertions.assertThat(periodeListeUtenJustering).isNotNull() },
            Executable { Assertions.assertThat(periodeListeUtenJustering.size).isEqualTo(3) },
            Executable { Assertions.assertThat(periodeListeUtenJustering[0].datoFom).isEqualTo(LocalDate.parse("2017-01-01")) },
            Executable { Assertions.assertThat(periodeListeUtenJustering[0].datoTil).isEqualTo(LocalDate.parse("2018-01-01")) },
            Executable { Assertions.assertThat(periodeListeUtenJustering[1].datoFom).isEqualTo(LocalDate.parse("2018-01-01")) },
            Executable { Assertions.assertThat(periodeListeUtenJustering[1].datoTil).isEqualTo(LocalDate.parse("2019-01-01")) },
            Executable { Assertions.assertThat(periodeListeUtenJustering[2].datoFom).isEqualTo(LocalDate.parse("2019-01-01")) },
            Executable { Assertions.assertThat(periodeListeUtenJustering[2].datoTil).isEqualTo(LocalDate.parse("2020-01-01")) }
        )
    }
}
